Source/Flow/Private/FlowAsset.cpp

```diff
@@ -90,74 +124,124 @@ EDataValidationResult UFlowAsset::ValidateAsset(FFlowMessageLog& MessageLog)
+
+	if (!CanFlowNodeClassBeUsedByFlowAsset(*FlowNodeClass))
+	{
+		return false;
+	}
+
+	if (!CanFlowAssetUseFlowNodeClass(*FlowNodeClass))
+	{
+		return false;
+	}
+
+	// Confirm plugin reference restrictions are being respected
+	if (!CanFlowAssetReferenceFlowNode(*FlowNodeClass, OutOptionalFailureReason))
 	{
 		return false;
 	}
 
-	UFlowNode* NodeDefaults = FlowNodeClass->GetDefaultObject<UFlowNode>();
+	return true;
+}
+
+bool UFlowAsset::CanFlowNodeClassBeUsedByFlowAsset(const UClass& FlowNodeClass) const
+{
+	UFlowNode* NodeDefaults = FlowNodeClass.GetDefaultObject<UFlowNode>();
 
 	// UFlowNode class limits which UFlowAsset class can use it
+	for (const UClass* DeniedAssetClass : NodeDefaults->DeniedAssetClasses)
 	{
-		for (const UClass* DeniedAssetClass : NodeDefaults->DeniedAssetClasses)
+		if (DeniedAssetClass && GetClass()->IsChildOf(DeniedAssetClass))
 		{
-			if (DeniedAssetClass && GetClass()->IsChildOf(DeniedAssetClass))
-			{
-				return false;
-			}
+			return false;
 		}
+	}
 
-		if (NodeDefaults->AllowedAssetClasses.Num() > 0)
+	if (NodeDefaults->AllowedAssetClasses.Num() > 0)
+	{
+		bool bAllowedInAsset = false;
+		for (const UClass* AllowedAssetClass : NodeDefaults->AllowedAssetClasses)
 		{
-			bool bAllowedInAsset = false;
-			for (const UClass* AllowedAssetClass : NodeDefaults->AllowedAssetClasses)
+			if (AllowedAssetClass && GetClass()->IsChildOf(AllowedAssetClass))
 			{
-				if (AllowedAssetClass && GetClass()->IsChildOf(AllowedAssetClass))
-				{
-					bAllowedInAsset = true;
-					break;
-				}
-			}
-			if (!bAllowedInAsset)
-			{
-				return false;
+				bAllowedInAsset = true;
+				break;
 			}
 		}
+		if (!bAllowedInAsset)
+		{
+			return false;
+		}
 	}
 
+	return true;
+}
+
+bool UFlowAsset::CanFlowAssetUseFlowNodeClass(const UClass& FlowNodeClass) const
+{
 	// UFlowAsset class can limit which UFlowNode classes can be used
+	for (const UClass* DeniedNodeClass : DeniedNodeClasses)
 	{
-		for (const UClass* DeniedNodeClass : DeniedNodeClasses)
+		if (DeniedNodeClass && FlowNodeClass.IsChildOf(DeniedNodeClass))
 		{
-			if (DeniedNodeClass && FlowNodeClass->IsChildOf(DeniedNodeClass))
-			{
-				return false;
-			}
+			return false;
 		}
+	}
 
-		if (AllowedNodeClasses.Num() > 0)
+	if (AllowedNodeClasses.Num() > 0)
+	{
+		bool bAllowedInAsset = false;
+		for (const UClass* AllowedNodeClass : AllowedNodeClasses)
 		{
-			bool bAllowedInAsset = false;
-			for (const UClass* AllowedNodeClass : AllowedNodeClasses)
-			{
-				if (AllowedNodeClass && FlowNodeClass->IsChildOf(AllowedNodeClass))
-				{
-					bAllowedInAsset = true;
-					break;
-				}
-			}
-			if (!bAllowedInAsset)
+			if (AllowedNodeClass && FlowNodeClass.IsChildOf(AllowedNodeClass))
 			{
-				return false;
+				bAllowedInAsset = true;
+				break;
 			}
 		}
+		if (!bAllowedInAsset)
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool UFlowAsset::CanFlowAssetReferenceFlowNode(const UClass& FlowNodeClass, FText* OutOptionalFailureReason) const
```

Comment:
This is the new check if the node is accessible from the plugin context of the flow asset.